lib/jsdom/browser/Window.js

```diff
@@ -330,55 +327,136 @@ function Window(options) {
+
+    // TODO: implement timer nesting level behavior.
+
+    const methodContextProxy = methodContext._globalProxy;
+    const handle = previousHandle !== undefined ? previousHandle : ++latestTimerId;
+
+    function task() {
+      if (!listOfActiveTimers.has(handle)) {
+        return;
+      }
+
+      try {
+        if (typeof handler === "function") {
+          handler.apply(methodContextProxy, args);
+        } else if (window._runScripts === "") {
+          vm.runInContext(handler, window, { filename: window.location.href, displayErrors: false });
+        }
+      } catch (e) {
+        reportException(window, e, window.location.href);
+      }
+
+      if (repeat) {
+        timerInitializationSteps(handler, timeout, args, { methodContext, repeat: true, previousHandle: handle });
+      }
+    }
+
+    if (timeout < 0) {
+      timeout = 0;
+    }
+
+    const nodejsTimer = setTimeout(task, timeout);
+    listOfActiveTimers.set(handle, nodejsTimer);
+
+    return handle;
+  }
+
+  // https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#animation-frames
+
+  let animationFrameCallbackId = 0;
+  const mapOfAnimationFrameCallbacks = new Map();
+  let animationFrameNodejsInterval = null;
 
   if (this._pretendToBeVisual) {
-    this.requestAnimationFrame = fn => {
-      const timestamp = rawPerformance.now() - windowInitialized;
-      const fps = 1000 / 60;
-
-      return startTimer(
-        window,
-        setTimeout,
-        clearTimeout,
-        ++latestAnimationFrameCallbackId,
-        fn,
-        fps,
-        animationFrameCallbacks,
-        [timestamp]
-      );
+    this.requestAnimationFrame = function (callback) {
+      callback = webIDLConversions.Function(callback);
+
+      const handle = ++animationFrameCallbackId;
+      mapOfAnimationFrameCallbacks.set(handle, callback);
+      return handle;
     };
-    this.cancelAnimationFrame = stopTimer.bind(this, animationFrameCallbacks);
-  }
 
-  this.__stopAllTimers = function () {
-    stopAllTimers(timers);
-    stopAllTimers(animationFrameCallbacks);
+    this.cancelAnimationFrame = function (handle) {
+      handle = webIDLConversions["unsigned long"](handle);
 
-    latestTimerId = 0;
-    latestAnimationFrameCallbackId = 0;
+      mapOfAnimationFrameCallbacks.delete(handle);
+    };
 
-    timers = Object.create(null);
-    animationFrameCallbacks = Object.create(null);
-  };
+    function runAnimationFrameCallbacks(now) {
+      const callbackHandles = mapOfAnimationFrameCallbacks.keys();
+      for (const handle of callbackHandles) {
+        if (mapOfAnimationFrameCallbacks.has(handle)) {
+          const callback = mapOfAnimationFrameCallbacks.get(handle);
+          mapOfAnimationFrameCallbacks.delete(handle);
+          try {
+            callback(now);
+          } catch (e) {
+            reportException(window, e, window.location.href);
+          }
+        }
+      }
+    }
+
+    animationFrameNodejsInterval = setInterval(() => {
```

Comment:
Although this is a lot simpler and closer to the spec, it probably wouldn't be appreciated by our consumers since it would mean you need to call `window.close()` on any `pretendToBeVisual` `JSDOM`, even with no `requestAnimationFrame()` calls in play. So this should probably change.

Reply:
Yeah, just pushed a commit to do that